pkg/operator/ceph/pool/controller.go

```diff
@@ -18,155 +18,219 @@ limitations under the License.
+	waitForRequeueIfNoCephCluster = reconcile.Result{Requeue: true, RequeueAfter: 1 * time.Minute}
+	// waitForRequeueIfOperatorIsNotReady waits for the operator to start
+	waitForRequeueIfOperatorIsNotReady = reconcile.Result{Requeue: true, RequeueAfter: 10 * time.Second}
+)
 
-// PoolController represents a controller object for pool custom resources
-type PoolController struct {
-	context     *clusterd.Context
-	clusterSpec *cephv1.ClusterSpec
+// ReconcileCephBlockPool reconciles a CephBlockPool object
+type ReconcileCephBlockPool struct {
+	client  client.Client
+	scheme  *runtime.Scheme
+	context *clusterd.Context
 }
 
-// NewPoolController create controller for watching pool custom resources created
-func NewPoolController(context *clusterd.Context, clusterSpec *cephv1.ClusterSpec) *PoolController {
-	return &PoolController{
-		context:     context,
-		clusterSpec: clusterSpec,
-	}
+// Add creates a new CephBlockPool Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, context *clusterd.Context) error {
+	return add(mgr, newReconciler(mgr, context))
 }
 
-// Watch watches for instances of Pool custom resources and acts on them
-func (c *PoolController) StartWatch(namespace string, stopCh chan struct{}) {
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, context *clusterd.Context) reconcile.Reconciler {
+	// Add the cephv1 scheme to the manager scheme so that the controller knows about it
+	mgrScheme := mgr.GetScheme()
+	cephv1.AddToScheme(mgr.GetScheme())
 
-	resourceHandlerFuncs := cache.ResourceEventHandlerFuncs{
-		AddFunc:    c.onAdd,
-		UpdateFunc: c.onUpdate,
-		DeleteFunc: c.onDelete,
+	return &ReconcileCephBlockPool{
+		client:  mgr.GetClient(),
+		scheme:  mgrScheme,
+		context: context,
 	}
-
-	logger.Infof("start watching pools in namespace %q", namespace)
-	go k8sutil.WatchCR(PoolResource, namespace, resourceHandlerFuncs, c.context.RookClientset.CephV1().RESTClient(), &cephv1.CephBlockPool{}, stopCh)
 }
 
-func (c *PoolController) onAdd(obj interface{}) {
-	if c.clusterSpec.External.Enable && c.clusterSpec.CephVersion.Image == "" {
-		logger.Warningf("Creating pools for an external ceph cluster is disabled because no Ceph image is specified")
-		return
-	}
-
-	pool, err := getPoolObject(obj)
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
 	if err != nil {
-		logger.Errorf("failed to get pool object. %v", err)
-		return
+		return err
 	}
-	updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.ProcessingStatus, c.context)
-	err = createPool(c.context, pool)
+
+	// Watch for changes on the CephBlockPool CRD object
+	err = c.Watch(&source.Kind{Type: &cephv1.CephBlockPool{}}, &handler.EnqueueRequestForObject{})
```

Comment:
> Is this watching all namespaces? If so:
> 
> * Is there a way to restrict it to just watch a certain namespace? Then we could follow the same pattern of only starting the child CR watchers after the cluster is confirmed as ready.

We currently rely on to watch a single or any namespace: https://github.com/rook/rook/blob/2b5aa9a4133aa5ec56345edf15887889cce4f6fa/pkg/operator/ceph/cr_manager.go#L33

> * If a pool CR gets created in the wrong namespace, the operator will keep printing messages that it can't be reconciled. Not a big deal, but just wanted to confirm.

No sure if the "wrong" namespace is a thing because we rely on the namespace from the CR object.

Reply:
I see, we will be watching either all namespaces or only the operator namespace, according to the operator settings.